### ISSUE TITLE
Wrong parsing of strings starting from 'a'

### ISSUE THREAD

**Issue by arsen91** [moment/moment#2986]
Hello. I found this issue:

`moment('adsfasdffgdgw', 'hh:mma').format('HH:mm')`

gives "00:00"

Don't know why it happens, but it happens with every string starting from 'a' letter.

**Comment by maggiepint** [comment]
What is happening here is that the parser is picking up on the letter 'a' as the meridiem indicator (AM/PM indicator) based on the 'a' token being present in the format, and thus interpreting that we have valid data.
You'll notice that the following results in invalid date (as expected):
```js
moment('adsfasdffgdgw', 'hh:mm').format('HH:mm')
```

It should be the case that if the only token that is parsed is the meridiem indicator, the date is invalid.

If you need to have this string be invalid in the interim, it might be beneficial to use strict parsing:

```js
moment('adsfasdffgdgw', 'hh:mma', true).format('HH:mm')
```

Note that this means that your date must exactly match the format provided in order to be considered valid.

**Comment by arsen91** [comment]
This is not good :)
I have `moment(time, 'hh:mma').format('HH:mm') ` and time variable can be either in 24 hours or in 12 hours format. And it formats correct in both cases. However when I use `true` as third param, as you said, moment expects date to

> exactly match the format provided

Ok, will need some more code to figure out the format.. Thanks

**Comment by maggiepint** [comment]
Why not just use multiple formats?

```js
moment('12:00', ['hh:mma', 'hh:mm'], true).format()
"2016-02-25T12:00:00-06:00"
moment('12:00a', ['hh:mma', 'hh:mm'], true).format()
"2016-02-25T00:00:00-06:00"
moment('12:00a', ['hh:mma', 'HH:mm'], true).format()
"2016-02-25T00:00:00-06:00"
moment('11:00p', ['hh:mma', 'HH:mm'], true).format()
"2016-02-25T23:00:00-06:00"
moment('23:00', ['hh:mma', 'HH:mm'], true).format()
"2016-02-25T23:00:00-06:00"
moment('asdkgjhlasdfg', ['hh:mma', 'HH:mm'], true).format()
"Invalid date"
```

**Comment by mj1856** [comment]
We've seen this before, in #2687 and similarly in #2956 (and others, IIRC).  However we've closed it as by design, with strict mode being the prefered solution.

We even call it out in the [docs](http://momentjs.com/docs/#/parsing/string-format/):

> Moment's parser is very forgiving, and this can lead to undesired behavior.
> ```javascript
> moment('I am spartacus', 'h:hh A').isValid();     //true - the 'am' matches the 'A' flag.
> ```

Perhaps is should be reevaluated.  Even in non-strict mode, it doesn't seem right to match solely on meridiem, or month, or really even on a single digit.  IMHO, it's reasonable to consider it a mismatch when a format token string is present and not all elements are present.  In other words:

```javascript
moment('I am spartacus', 'h:mm A').isValid(); // false
moment('I am 1:23 AM spartacus', 'h:mm A').isValid();  // true 
moment('I am 1:23 AM spartacus', 'h:mm A', true).isValid();  // false - strict mode
```

I'd also argue that the example in the docs should throw an error, as there are two separate hour tokens `h` and `hh`, so a valid input is never possible.  I'll raise a separate issue for that.

**Comment by arsen91** [comment]
> IMHO, it's reasonable to consider it a mismatch when a format token string is present and not all elements are present

Absolutely agree. Thank you!

**Comment by mj1856** [comment]
Should be resolved when #3045 is merged.  Will track there.  Thanks!
